src/spec.c

```diff
@@ -1799,25 +1856,23 @@ static IndexesScanner *IndexesScanner_New(IndexSpec *spec) {
 void IndexesScanner_Free(IndexesScanner *scanner) {
   if (global_spec_scanner == scanner) {
     global_spec_scanner = NULL;
-  } else if (!scanner->cancelled) {
-    if (scanner->spec && scanner->spec->scanner == scanner) {
-      scanner->spec->scanner = NULL;
-      scanner->spec->scan_in_progress = false;
+  } else {
+    StrongRef tmp = WeakRef_Promote(scanner->spec_ref);
+    IndexSpec *spec = StrongRef_Get(tmp);
+    if (spec) {
+      if (spec->scanner == scanner) {
+        spec->scanner = NULL;
+        spec->scan_in_progress = false;
+      }
+      StrongRef_Release(tmp);
     }
+    WeakRef_Release(scanner->spec_ref);
   }
-
+  if (scanner->spec_name) rm_free(scanner->spec_name);
   rm_free(scanner);
 }
 
-void IndexesScanner_Cancel(IndexesScanner *scanner, bool still_in_progress) {
-  if (scanner->cancelled) {
-    return;
-  }
-  if (!scanner->global && scanner->spec) {
-    scanner->spec->scan_in_progress = still_in_progress;
-    scanner->spec->scanner = NULL;
-    scanner->spec = NULL;
-  }
+void IndexesScanner_Cancel(IndexesScanner *scanner) {
```

Comment:
Why would we want to cancel? It will be cancelled because we will not be able to promote  the weak ref no?

Reply:
It might get canceled when a new scanner replaces a running one, without invalidating the spec. we can "trick" the scanner and replace its reference with an invalid one instead

Reply:
Ohh, in case of alter the index? :+1:

Reply:
Exactly